### ISSUE TITLE
XmlNode explore = runner.Explore(TestFilter.Empty);  //cause lock Engine 3.17

### ISSUE THREAD

**Issue by PatrickLeGarrec** [nunit/nunit-console#1425]
Hi :)              

```
CopyFileWithoutLock(sourceFilePath, destinationFilePath);   // my FileStream implementation

ITestEngine engine = TestEngineActivator.CreateInstance();
TestPackage package = new TestPackage(destinationFilePath);
ITestRunner runner = engine.GetRunner(package);
XmlNode explore = runner.Explore(TestFilter.Empty);
```
The file copy is ok and the exploration **works**.

then i redo a copy from the same file to the same destination

`CopyFileWithoutLock(sourceFilePath, destinationFilePath);`

 ==> Failed ==> " **is being used by another process** " The file is **locked**.

Then i redo the operation without exploring, with x2  successfull copy operations **without any lock**.

```
CopyFileWithoutLock(sourceFilePath, destinationFilePath);

ITestEngine engine = TestEngineActivator.CreateInstance();
TestPackage package = new TestPackage(destinationFilePath);
ITestRunner runner = engine.GetRunner(package);

CopyFileWithoutLock(sourceFilePath, destinationFilePath);

```
The file copy is ok ==>  no lock.

It seems that the engine.GetRunner cause lock, even if i dispose() both engine and runner.
or runner.Explore.

Any clue ?

**Comment by CharliePoole** [comment]
Most of the methods defined in `ITestRunner` , including `Explore`, force a call to `ItestRunner.Load`, which (obviously) loads the test assemblies specified in your `TestPackage`. This has always implied a guarantee to the caller that the loaded test assemblies will __remain__ loaded until they are unloaded or until the process exits.

This behavior is required by any runner, e.g. a gui, which displays a representation of the tests and allows the user to select individual tests for running or view information about such tests. So this has to be "by design" or "won't fix".

You could try actually unloading the explored tests, which may work or not depending on what else your application is doing with them. As an example of the problems which may arise, test cases randomly generated by nunit will not be the same across an unload.

**Comment by CharliePoole** [comment]
If you encounter a bug using Unload, we can reopen this.

**Comment by PatrickLeGarrec** [comment]
You describe perfectly what i needed.

Unload does not work too but it is normal.
I'm trying to do something that's not possible that way.

I create a small demo project to check the capabilities of the engine.

I was thinking about your test-centric GUI. 
You load Nunit dll and discover the tests as usual, and run them.
I do the same it works nice.

but you need to explore the unknown .dll just loaded, and the user may load another dll, etc...
I mean you don't have any lock rather than i have :)

Thanks anyway

**Comment by CharliePoole** [comment]
Can you share a minimal repro? If so, I'll take another look. While the tests must be kept loaded, that does not imply that the file from which they were loaded should be locked. In the V2 and TC GUIs, for example, that file is not locked but is watched to see if it changes. Since V3, NUnit itself has not had a GUI, so some code that is not exercised regularly can end up being removed.

**Comment by PatrickLeGarrec** [comment]
Here is :  https://github.com/PatrickLeGarrec/DemoNUnitEngine
Thanks in advance for your help.

**Comment by PatrickLeGarrec** [comment]
We just fixed it.  The dependency on test.dll is mandatory. So the file is locked normaly. 
so we copied and renamed the test.dll and explore the renamed file. 
it works and no lock persist anymore because there is no dependency on this new file.

The dll **must be** a renamed copy of the dependency  (detail of the renamed file is still test.dll) to work properly.
We understood that nUnit engine did not need to do with that. We were wrong.

**Comment by CharliePoole** [comment]
IMO it was a mistake to require tests run under test explorer to _reference_ the appropriate adapter. In the design of NUnit - and of other X-unit style frameorks  written in the same era - a test _should have no information_ about how or by which runner it will be executed. Unfortunately, that ship has sailed.

I'll look at your repro after I finish a few other things. As you state, there's no reason the file should be locked.

**Comment by PatrickLeGarrec** [comment]
No problem Charlie, take your time, it's holiday time here so :)
I m still investigating and found some clues.  I will try to do not flood the thread. 
I note i missed some packages with some earlier versions. actually the test.dll is explored well with older libraries, not with the last ones. But yes, the file is lock. and i m investigating...

**Comment by PatrickLeGarrec** [comment]
I update the repo with a better version. The package combo is old but the exploration now works with the same code.
Earliers versions  (NUnit / Engine / Adapter) make the Test.dll not found by the engine.

**Comment by PatrickLeGarrec** [comment]
Well, i have ran deeper into the engine code and found that the issue would be caused by the Assembly loading.
It is known that Assembly.Load and LoadFrom cause lock, and it is better to load in memory with File.ReadAllBytes instead.
but it may occurs some memory leak. I have tried to change that but the lock is still there. The parts of code are under NETSTANDARD2_0. It does nothing.

Also, i see AssemblyLoadContext and CustomAssemblyLoadContext are called under NETCOREAPP3_1_OR_GREATER and the code is called here too.
According to this https://learn.microsoft.com/en-us/dotnet/standard/assembly/unloadability ,  it seems that the context is never unload(). i maybe wrong and need to investigate more.

I m sure that the lock is due to this, because if i neutralize the call to LoadResult from MasterTestRunner.cs  Explore

```
  public XmlNode Explore(TestFilter filter)
  {
   //   LoadResult = PrepareResult(GetEngineRunner().Explore(filter))
   //       .MakeTestRunResult(TestPackage);
      XmlNode node = null;
      return node; 
    //  return LoadResult.Xml;
  }
```

there is no lock on the assembly.

**Comment by PatrickLeGarrec** [comment]
some news about engine 3.17

If i use the nuget package the test.dll is not found.

If i make a dependency directly on the engine 3.17 project (see below), the test.dll is found and the xml is ok.

![317](https://github.com/nunit/nunit-console/assets/89050431/1ee324f9-14cc-475c-a473-c89305aaf7b7)

Lock is still there, i will now check under this version

**Comment by CharliePoole** [comment]
You can also look at the latest 3.18 dev build on myget, currently https://www.myget.org/feed/nunit/package/nuget/NUnit.Engine/3.18.0-dev00058

**Comment by PatrickLeGarrec** [comment]
3.18.0-dev00058    => NUnit.Engine.NUnitEngineException: An exception occurred in the driver while loading tests.
 ---> System.IO.FileNotFoundException: Could not load file or assembly 'Test, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.

![318](https://github.com/nunit/nunit-console/assets/89050431/0e9fbc07-dca2-4646-9b1b-f49420c54ca0)

Same behavior as 3.17  => file not found

lock is still there in any version, will continue to investigate on 3.17 source code

**Comment by PatrickLeGarrec** [comment]
i just remove   test-centric.engine.metadata .
Now it works, dll is read.

the test centric metadat is provided with the 3.17 package also. and fail to read the test.
With the source code, there is no tc metadata and it works.

So the idea to remove the tc-metadata dependency fix the issue for me.

![318_2](https://github.com/nunit/nunit-console/assets/89050431/9e0c8e0e-f678-4ee3-83da-c3d015ad12ab)

**Comment by CharliePoole** [comment]
We're talking about the 3.17 __engine__ package, right?

So, the engine makes use of TC metadata to examine assemblies and decide how to run them. In order for the engine to work completely, it needs to be present. Are you sure there is not a copy around from some other source?

**Comment by PatrickLeGarrec** [comment]
3.17 and  3.18.
i will check.

**Comment by CharliePoole** [comment]
The metadata is a build dependency but the dll is bundled with the package. We might want to change that.

**Comment by PatrickLeGarrec** [comment]
Well, you were right.  
When i have used the engine source project as dependency, the tc-metadata 1.7.1 was in the dir, so i was wrong with my Analysis

Thanks for the information about the role of tc metadata dll.

about 3.17, what i notice is when i run the package, test.dll is not read.
but if i make a dependency on test.dll, it works. i think i know why now.

in the Test debug dir (NUnit3TestAdapter 4.5.0), tc metadata dll  is 1.7.1 30/10/2021 20:40  (EU datetime)

in the console debug dir with 3.17 engine package, tc metadata dll is  2.0.0. 05/03/2023  18:31  

Now, i found something :

With the 3.17 source project, the tc-metadata dll is 1.7.1 30/10/2021 and works.
the 3.17 package is 2.0.0 05/03/2023, and fail to read.

**Comment by CharliePoole** [comment]
TC Metadata version 2 has breaking changes, so if  somebody (I think before me) updated 3.17 to use it they may not have correctly updated how they are doing it. Seems doubtful, but possible. The breaking change is that metadata 1.x uses the Mono.Cecil namespace, which caused conflicts with some user code. The 2.x version uses TestCentric.Metadata. There is also a 3.x, with another breaking change: the name of the assembly.

@PatrickLeGarrec It seems that you like digging deep into obscure problems. Would you like to help on this project? We can talk about it offline via email... mine is listed on my profile.

**Comment by PatrickLeGarrec** [comment]
Yes, i see the error about Mono.Cecil. Ok there is a dll version issue, i happen so often... 

I can check the last version of the engine now, to find what goes wrong with the lock. 
It is an Assembly unload issue for sure, domain context concerns.

I develop a nUnit runner for our customer and strategic offer, which have special capabilities around playwright and E2E business testing with extended capabilities, far from cuncumber and robot framework i don't llike.

It is a very special use case and i encourage other dev team to use Nunit, i help to build some connector with their tool, Altassian addon's and others. Nunit is powerfull and work nice with multi threading. The Engine is good (but the lock still offense me)

That s why i focus on NUnit because the framework, engine and stuff are excellent and i m a C# coder.

Yes we can talk offline, I m always aware about the future of NUnit and overall Microsoft activities.

**Comment by PatrickLeGarrec** [comment]
I think i m off for today and will dig into the 3.17 to find what s wrong with the lock.
It is a problem for me if my tool can't load dll on the fly and do some File operation due to Assembly lock.

**Comment by PatrickLeGarrec** [comment]
Hi Charlie, some fresh and good news today. 

I found the lock origin, have a solution for it, and found a strange thing too. I explain :

TestAssemblyLoadContext.cs

=> TestAssemblyLoadContext is not collectible
according to https://learn.microsoft.com/en-us/dotnet/standard/assembly/unloadability
and https://learn.microsoft.com/en-us/dotnet/standard/assembly/unloadability
It should be collectible.

In order to allow Unload() on the TestAssemblyLoadcontext and unLoad the assembly(ies), in our case the tests (=package). 

```
public class TestAssemblyLoadContext : AssemblyLoadContext
{
    publicTestAssemblyLoadContext() : base(isCollectible: true)
```

Also, the Assembly should be loaded into the context as a ByteArray, and disposable. 
Load and LoadFrom cause lock, always.

The following code is an example of a working Assembly, collectible LoadContext without creating any lock on dll 👍 

```
var alc = new TestAssemblyLoadContext();
Stream ms = new MemoryStream(File.ReadAllBytes(dllPath));
Assembly a = alc.LoadFromStream(ms);

// ...... doing our stuff here

alc.Unload();
ms.Dispose();

File.Delete(dllPath);     // <== NO LOCK

class TestAssemblyLoadContext : AssemblyLoadContext
{
    public TestAssemblyLoadContext() : base(isCollectible: true) {}

    protected override Assembly? Load(AssemblyName name)
    {
        return null;
    }
}
```
Stream ms = new MemoryStream(File.ReadAllBytes(dllPath));  is the key to replace the Assembly.Load and LoadFrom if you don't want to have some lock. But to avoid some memory leaks it s important to do not forget to dispose the stream.
But if you want to keep the control and secure with a dll (like the engine itself), it may be interesting to keep the classic lock way. But in our case, the Test.dll (package) should not be locked for any reason.

Now the next step is to implement the fix.

I think it should be into the runner.unload()  or dispose().
I found that engine.dispose() call the unload(), but.... do nothing :

TestAgentRunner.cs

```
  protected virtual void Dispose(bool disposing)
  {
      if (!_disposed)
      {
          if (disposing)
              Unload();

          _disposed = true;
      }
  }

public virtual void Unload() {  }

```

I don't know how the have access to TestAssemblyLoadContext  and the Assembly (test) from here in order to perform the needed releasing operations.

**Comment by CharliePoole** [comment]
@PatrickLeGarrec
It seems to me that the Load or LoadFrom ought to work as well, provided that we correctly handle any "Unloading" events. Also, I'd like to compare the handling of this in nunit-engine vs my test-centric engine, which is used in a GUI. I don't rule out a major change such as you describe, but it's substantial so I'd like to try lesser changes, at leasat for the upcoming 3.18 release. Version 4 is another matter, of course.

Would you like to work on doing a PR?

**Comment by CharliePoole** [comment]
So, after all this... what in your opinion is the minimum code needed to demonstrate the error? Load + Unload + Check for file locked? We could do that for both .NET Framework using an AppDomain and .NET Core using an AssemblyLoadContext.

**Comment by CharliePoole** [comment]
I created unit tests, which demonstrate that the assembly files are locked when loaded and stay locked after unload. This is actually consistent for both .NET Framework and .NET Core but does not match what I thought was the historical behavior of NUnit under the .NET Framework. So... something changed over the years. Tests are currently in a local branch on my machine.

**Comment by PatrickLeGarrec** [comment]
Well... after digging a lot, i think i have reached the center of the earth ;) and flood the sea but anyway, it worth the price of time because i learn how the engine works.

So, the story about TestAssemblyLoadContext and the collectible=true gives the capabilities to perform an unload of the AssemblyLoadContext. if collectible is false, you can't.
But by doing this, all the context with the others loaded assemblies are unload too. 
I think it s ok but at a very late stage, because a user can explore and perform some execution (run) of the tests, etc...
without purging the context. but it is an other story...may be i m wrong.

But anyway, the changes may occurs a design break in the code. I agree with you.

Then, i retry some code without worry about the collectible flag and the unload, and it makes the things simple.
I take the 3.17 code "vanilla" and just track for the lock. I found exactly where it is. and it is not what i was expected :

To help me i use a powertoy File Locksmith and finally there is a Pro, and a Con. I explain :

the following pic show the issue with basic code :

![error](https://github.com/nunit/nunit-console/assets/89050431/a156fc00-ffcf-40e2-8df8-80ad86cca6d6)


In the code, i found that 

The vanilla code make lock :

![Capture d’écran 2024-07-05 002948](https://github.com/nunit/nunit-console/assets/89050431/536ae855-7da5-46d0-bc16-0f3e08ea4803)

My code DO NOT make lock 

![Capture d’écran 2024-07-05 002811](https://github.com/nunit/nunit-console/assets/89050431/149f7abb-642c-4a08-9b7e-79e469f35377)

BUT there is a problem later :

![Capture d’écran 2024-07-05 002413](https://github.com/nunit/nunit-console/assets/89050431/b160de13-f87d-46d3-b117-b9bd928f9686)

```
NUnit.Engine.NUnitEngineException: An exception occurred in the driver while loading tests.
 ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.ArgumentException: Argument name must not be the empty string (Parameter 'name')
   at NUnit.Framework.Guard.<ArgumentNotNullOrEmpty>g__ThrowArgumentNotNullOrEmpty|1_0(String name)
   at NUnit.Framework.Guard.ArgumentNotNullOrEmpty(String value, String name)
   at NUnit.Framework.Internal.Test..ctor(String pathName, String name, ITypeInfo typeInfo, IMethodInfo method)
   at NUnit.Framework.Internal.Test..ctor(String name)
   at NUnit.Framework.Internal.TestSuite..ctor(String name)
   at NUnit.Framework.Internal.TestAssembly..ctor(String assemblyNameOrPath)
   at NUnit.Framework.Api.DefaultTestAssemblyBuilder.Build(Assembly assembly, String assemblyNameOrPath, IDictionary`2 options)
   at NUnit.Framework.Api.DefaultTestAssemblyBuilder.Build(Assembly assembly, IDictionary`2 options)
   at NUnit.Framework.Api.NUnitTestAssemblyRunner.<>c__DisplayClass34_0.<Load>b__0()
   at NUnit.Framework.Api.NUnitTestAssemblyRunner.WrapInNUnitCallContext[T](Func`1 function)
   at NUnit.Framework.Api.NUnitTestAssemblyRunner.Load(Assembly assembly, IDictionary`2 settings)
   at NUnit.Framework.Api.FrameworkController.LoadTests()
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
   --- End of inner exception stack trace ---
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at NUnit.Engine.Drivers.NUnitNetCore31Driver.ExecuteMethod(MethodInfo method, Object[] args) in C:\Repositories.Net\nunit-console-3.17.0\src\NUnitEngine\nunit.engine.core\Drivers\NUnitNetCore31Driver.cs:line 213
   at NUnit.Engine.Drivers.NUnitNetCore31Driver.ExecuteMethod(String methodName, Object[] args) in C:\Repositories.Net\nunit-console-3.17.0\src\NUnitEngine\nunit.engine.core\Drivers\NUnitNetCore31Driver.cs:line 195
   at NUnit.Engine.Drivers.NUnitNetCore31Driver.Load(String assemblyPath, IDictionary`2 settings) in C:\Repositories.Net\nunit-console-3.17.0\src\NUnitEngine\nunit.engine.core\Drivers\NUnitNetCore31Driver.cs:line 110
   at NUnit.Engine.Runners.TestAgentRunner.LoadDriver(IFrameworkDriver driver, String testFile, TestPackage subPackage) in C:\Repositories.Net\nunit-console-3.17.0\src\NUnitEngine\nunit.engine.core\Runners\TestAgentRunner.cs:line 149
   --- End of inner exception stack trace ---
   at NUnit.Engine.Runners.TestAgentRunner.LoadDriver(IFrameworkDriver driver, String testFile, TestPackage subPackage) in C:\Repositories.Net\nunit-console-3.17.0\src\NUnitEngine\nunit.engine.core\Runners\TestAgentRunner.cs:line 153
   at NUnit.Engine.Runners.TestAgentRunner.Load() in C:\Repositories.Net\nunit-console-3.17.0\src\NUnitEngine\nunit.engine.core\Runners\TestAgentRunner.cs:line 134
   at NUnit.Engine.Runners.TestAgentRunner.EnsurePackageIsLoaded() in C:\Repositories.Net\nunit-console-3.17.0\src\NUnitEngine\nunit.engine.core\Runners\TestAgentRunner.cs:line 265
   at NUnit.Engine.Runners.TestAgentRunner.Explore(TestFilter filter) in C:\Repositories.Net\nunit-console-3.17.0\src\NUnitEngine\nunit.engine.core\Runners\TestAgentRunner.cs:line 74
   at NUnit.Engine.Runners.MasterTestRunner.Explore(TestFilter filter) in C:\Repositories.Net\nunit-console-3.17.0\src\NUnitEngine\nunit.engine\Runners\MasterTestRunner.cs:line 214
   at NUnitWrappers.NUnitExplorer.Explore() in C:\Repositories.Net\DemoNUnitEngine-main-3-17_simple\ConsoleApp\NUnitExplorer.cs:line 33
```

Parameter 'name' is missing during reflection.

We know that doing this

```
Stream ms = new MemoryStream(File.ReadAllBytes(assemblyPath));                      
_dependencyContext = DependencyContext.Load(loadContext.LoadFromStream(ms)); 
```

We loose information, (location and codebase)... because this is just a flow of byte in memory.
And reflection seems to not enjoy ...

To finish, i think there is very very few change to bring to the code to make the lock away.
But, the reflection issue is problematic,... and i don't have the knowledge with the engine.

a PR is not necessary... the goal is now to know if there are breakings change.

The lock on the file is problematic. We can find a solution, a non-sexy workaround with renaming and tmp file to purge at the beginning of a session, just before new lock appairs,  but it is not the best, just not impossible.

**Comment by CharliePoole** [comment]
Separate problems exist for .NET Framework and .NET Core. In the .NET Framework case, we unload the test by unloading the AppDomain. This made sense in the early years of NUnit. BUT, there is a time delay for threads to finish, so we create a thread, which will wait for the domain to be unloaded, and simply return. This works in the nunit3-console context, because we only unload when we are ending the process. It _usually_ works in a GUI, because the user does not click to open a new file very quickly.

**Comment by PatrickLeGarrec** [comment]
Yes i understand, you can try this code
```

class TestAssemblyLoadContext : AssemblyLoadContext
{
    public TestAssemblyLoadContext() : base(isCollectible: true) {}
     ......... some code
}
```

For .NetFramework, it s more tricky yes.

**Comment by PatrickLeGarrec** [comment]
you can try this safe code
```

WeakReference testAWeakRef;
var alc = new TestAssemblyLoadContext();

// the no-lock version
Stream ms = new MemoryStream(File.ReadAllBytes(dllPath));
Assembly a = alc.LoadFromStream(ms);
ms.Dispose();

// the lock version
Assembly a = alc.Load(dllPath);

testAlcWeakRef = new WeakReference(alc, trackResurrection: true);

// and when you decide to purge the context :

alc.Unload();
for (int i = 0; testAWeakRef.IsAlive && testAlcWeakRef.IsAlive && (i< 10); i++)
{
    GC.Collect();
    GC.WaitForPendingFinalizers();
}


// and the flag to the constructor to do not forget
    class TestAssemblyLoadContext : AssemblyLoadContext
    {
        public TestAssemblyLoadContext() : base(isCollectible: true) {}   // <= with the flag, give unload() available.
```

yes for .Net Framework it is more tricky.

**Comment by PatrickLeGarrec** [comment]
What i have discovered is that the lock persist even after a safe Unload() of the context with all the resources disposed.

**Comment by PatrickLeGarrec** [comment]
If we know why the name is missing during the reflection

```
 ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.ArgumentException: Argument name must not be the empty string (Parameter 'name')
   at NUnit.Framework.Guard.<ArgumentNotNullOrEmpty>g__ThrowArgumentNotNullOrEmpty|1_0(String name)
   at NUnit.Framework.Guard.ArgumentNotNullOrEmpty(String value, String name)
```

And if we found a valid workaround, the lock may be fixed by transfering data in memory.  By changing how the dll is loaded into the loadcontext. LoadFromStream is respectable because it is part of the framework and can be used this way.

**Comment by PatrickLeGarrec** [comment]
if i compare the Assembly returned by LoadFromAssemblyPath and  LoadFromStream, they are the same, no difference

```
  Assembly asm = loadContext.LoadFromAssemblyPath(assemblyPath);

  Stream ms = new MemoryStream(File.ReadAllBytes(assemblyPath));
  Assembly asm2 = loadContext.LoadFromStream(ms);

  if (asm == asm2) { Console.WriteLine("no difference"); }
```

**Comment by CharliePoole** [comment]
So does this workaround resolve your use case? If so, then I'd like to consider this approach for V4 rather than 3.18 - and for TestCentric possibly, as well.

**Comment by PatrickLeGarrec** [comment]
We will use a workaround based on renaming and tmp file to purge at the begining of each process session. It is non-sexy and a bad design but it will do the trick. In the case of a testing framework with an engine like this one, very potent (our choice by default), we wait for more. The fact to have a lock on the files is problematic when the platform is shared between desktop ui app, and an automatic runner designed to load and execute grappes of tests on test farms as fast as possible, workflow based.

So we are aware about the evolution of the engine. NUnit is pretty used now as an E2E web/api testing framework, connected to tierce party plugins of worldwide companies, and not only for vs unit testing. The engine is excellent, and how the tests are designed too. But for industrial testing solution, a lock on a file is a limitation.

About the version, we will wait, no problem with that :)

**Comment by CharliePoole** [comment]
@PatrickLeGarrec has surfaced an interesting problem and proposed a new way to load tests. While this approach is hidden from users, it could have side-effects and I prefer to avoid it for now and it only impacts those who make use of the engine directly rather than through the console.  As a result, I'm re-categorizing this as a new feature and targeting the version 4 release.

Summary of earlier discussion: file locks may be avoided by reading the assembly into a memory stream and loading from that stream.  @PatrickLeGarrec demonstrates how to do this for .NET Core but it should also be adaptable to use in .NET Framework. 

A similar approach is already used in `TestCentric.Metadata`, which we utilize to examine assemblies without actually loading them.

**Comment by PatrickLeGarrec** [comment]
Here is the workaround to avoid the lock issue :

Desktop UI
(the desktop app use an autoload system for quality of life convenience)

```
private void frmBaseLayout_Load(object sender, EventArgs e)
{
    // Purge all the previously locked tests files at startup
    foreach (string strFile in Directory.GetFiles(AppDomain.CurrentDomain.BaseDirectory, searchPattern: "*_tmp.dll"))
    {
        if (Properties.Settings.Default.IsAutoLoad)
        {   // exclude the file to autoload
            if (strFile != Properties.Settings.Default.LastDllLoaded)
                File.Delete(strFile);
        }
        else File.Delete(strFile);
    } 
}



And when i load the tests from the Desktop GUI :

// ######################
// ### MANUAL LOADING ###
// ######################

public void UI_LoadAssembly()
{
    try
    {
        if (Handler.Instance.UcLoadIteration.openFileDialogAssembly.ShowDialog() == DialogResult.OK)
        {
            // generate a randomized temp file. get around the lock issue. see frmBaseLayout load() for the deleting routine.
            Random rnd = new();
            int randomIntInRange = rnd.Next(1000000);
            string fileName = Handler.Instance.UcLoadIteration.openFileDialogAssembly.SafeFileName;
            string fileNameNoExtension = fileName.Replace(".dll", "");
            string renamedFile = fileNameNoExtension + "." + randomIntInRange + "_tmp.dll";
            File.Copy(Handler.Instance.UcLoadIteration.openFileDialogAssembly.FileName, renamedFile, true);  // copy the new file
            Properties.Settings.Default.LastDllLoaded = renamedFile; 

            if (Handler.Instance.Fixture.ExploreFixtureAssembly(renamedFile))  // explore the new file with engine
            {   ..... some code.....
```

For information purpose.

Have a nice day.

**Comment by CharliePoole** [comment]
@nunit/engine-team @nunit/core-team 

@patlgrc initiated this issue due to a problem in the use of engine by a  third-party runner. It also affects my own TestCentric GUI. In the case of TestCentric, we have resolved this by wrapping the engine's runner with our own class, which creates a new runner for each explore or run operation and disposes it immediately after the operation is complete.

Here's what I'm thinking of doing. I'd appreciate your feedback...

1. I can't think of any situation where the console runner itself needs this behavior. In fact, for very large runs, creating each agent twice, once on load and once on execution, would be detrimental to performance. So I don't think this should become the default behavior of the engine. Neither do I see a need for a console option to activate it.
2. I do think it would be worthwhile to build this behavior into the engine as an option. Code we have in TestCentric serves as a starting point although it would need to be extended a bit, since TC only runs tests asynchronously.
3. The feature could be activated through a package setting or by use of a new optional parameter in the call to `GetRunner`. I lean toward the second approach.

**Comment by CharliePoole** [comment]
The engine currently has a `PackageSetting`, called `DisposeRunners`, which would appear to address this problem, at least partially.

Currently, the setting is only consumed by `AggregatingTestRunner`, so it has no effect when running a single assembly. When it is used, the individual runners are disposed immediately after the tests run is complete for each assembly. IIRC, this was added to improve efficiency when a large number of assemblies were being run... particularly when the number of assemblies was substantially greater than the number of agents to be used.

My initial thinking is to extend the semantics of this setting for V4...
1. `ProcessRunner` rather than `AggregatingTestRunner` should dispose of individual engine runners on completion of a run, so that single assembly and multiple assembly runs work the same way.
2. Disposal should occur after an `Explore` operation as well as a test run.
3. `ProcessRunner` should create the engine runner at the start of each run or explore operation.
4. The console runner currently sets `DisposeRunners` to true. We may need to rethink this if the semantics change. 

@nunit/engine-team Thoughts?

**Comment by manfred-brands** [comment]
@CharliePoole If in step 3 engine runners are created, why would dispose be a setting? If engines are disposable they should always be disposed or instances kept for reuse and then disposed at the end.

What would not disposing give you?
Note that any handles would eventually be closed by the Garbage Collector.

The only think of keeping the runner and client assembly loaded is so that one can run individual tests without the overhead of explore and load. If that is the use case, then a UI needs to hold on to an instance and have a manual unload functionality.
For the console runner the above doesn't hold as it would be a fresh start for each run.

I haven't looked at any code as I'm AFK.

**Comment by CharliePoole** [comment]
@manfred-brands 

This is all in the context of third-party runners that display the tests and allow multiple executions. The Console Runner of course only allows you to execute once and then terminates.

> If in step 3 engine runners are created, why would dispose be a setting? If engines are disposable they should always be disposed or instances kept for reuse and then disposed at the end.

We're talking runners here, specifically engine runners (ProcessRunner, etc.) not the engine and not the top-level `MasterTestRunner`.
I didn't explain that as well as I should have. Originally, there was no `DisposeRunners` setting. Stuff was disposed when the runner terminated. I believe that the setting was added to the V3 engine for more efficient use of memory. Possibly also for the adapter. 

> What would not disposing give you?
Note that any handles would eventually be closed by the Garbage Collector.

Without the setting, every engine runner you create along with the associated process would remain in memory even after that assembly had executed. The test assembly file would continue to be open. Imagine a test with 100 assemblies and the number of agents set to 20. In that case, only 20 processes would execute at a time but eventually all 100 would be in memory. Using the setting, no more than 20 processes would be in memory at one time. Under V3, we decided to turn it on all the time for the console runner.

> The only think of keeping the runner and client assembly loaded is so that one can run individual tests without the overhead of explore and load. If that is the use case, then a UI needs to hold on to an instance and have a manual unload functionality.
For the console runner the above doesn't hold as it would be a fresh start for each run.

That's exactly the use case for which this is intended. It applies equally to running individual tests, entire assemblies or the full test run. I'd prefer to implement the feature without affecting the console runner while still making the setting available for others to use, including myself for the TC GUI. I thought of several ways to do this, including a new API, a new setting or simply closing the file once all the info was in memory. But then I realized we already support a setting that does part of the job... i.e. it releases the engine runners but only under the `AggregateTestRunner` used for multiple assemblies. `ProcessRunner` OTOH is used for every test assembly whether it's run by itself or as one of many assemblies. So I'd remove the functionality from ATR and put it into PR. I'd also extend the same functionality so that it applies to Explore in addition to Run.

Regarding reuse of runners, I think it's a desirable goal since a lot of processing goes into determining all the characteristics of each engine runner, what settings to use, etc. But all that work also makes it pretty difficult to guarantee a saved runner is capable of running a particular package, especially since the package may have changed between runs. I'd rather save that for another time. For now, I think that most of the code needed to recreate an agent and runner is already in place but I'll need to test to be sure.


FYI, if we _didn't_ already have the `DisposeRunners` setting, I'd probably just add an optional parameter to `GetRunner` to tell it to dispose runners after each use.

**Comment by manfred-brands** [comment]
There is one thing that affects tests re-runnability: static initializers.
Those will be run only once, regardless.
We just fixed a bug in nunit's own tests where an semaphore was created in a static initializer and disposed in OneTimeTeardown.
It was fine until someone ran tests multiple times without reloading the assembly.

**Comment by CharliePoole** [comment]
Right. A runner such as the adapter or a GUI can point out errors like that. I think the engine uses statics in a safe manner but I think some tests could have similar problems.
